[improve][pip] PIP-425: fix problem that consumer or producer create failed when many nodes  in PulsarClient serviceUrl become unavailable

AuroraTwinkle opened the pull request:
…failed when many nodes  in PulsarClient serviceUrl become unavailable

<!--
### Contribution Checklist
  
  - PR title format should be *[type][component] summary*. For details, see *[Guideline - Pulsar PR Naming Convention](https://pulsar.apache.org/contribute/develop-semantic-title/)*. 

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

<!-- Either this PR fixes an issue, -->

Fixes #xyz

<!-- or this PR is one task of an issue -->

Main Issue: #xyz

<!-- If the PR belongs to a PIP, please add the PIP link here -->

PIP: #xyz 

<!-- Details of when a PIP is required and how the PIP process work, please see: https://github.com/apache/pulsar/blob/master/pip/README.md -->

### Motivation

<!-- Explain here the context, and why you're making that change. What is the problem you're trying to solve. -->

### Modifications

<!-- Describe the modifications you've done. -->

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [ ] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: <!-- ENTER URL HERE -->

<!--
After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.
-->

AuroraTwinkle commented:
> Some initial feedback

Updated!

lhotari commented on pip/pip-425.md:
```diff
@@ -0,0 +1,127 @@
+Implementation PR: [#24387](https://github.com/apache/pulsar/pull/24387)
+
+# Background knowledge
+In the current Pulsar client versions, consumer or producer creation is the responsibility of PulsarClient, we can build a PulsarClient by pass serviceUrl param. The serviceUrl param supports multiple endpoints with the syntax `pulsar://localhost1:6650,localhost2:6650,localhost3:6650,...` or `pulsar+ssl://localhost1:6650,localhost2:6650,localhost3:6650,...`. What happens if some of the endpoints  ...
```
drop "local" prefix in the examples to make them more generic. For pulsar+ssl, the example port should be 6651.

lhotari replied:
Could this be explained more directly? "Currently when an unavailable endpoint is used from the list of multiple endpoint, the client will fail to create a consumer or producer. " And also explaining whether this error is transient or persistent.

lhotari commented on pip/pip-425.md:
```diff
@@ -0,0 +1,127 @@
+
+# Motivation
+
+As [#22934 ](https://github.com/apache/pulsar/issues/22934) and [#22933](https://github.com/apache/pulsar/discussions/22933) mentioned, when most of the endpoints in serviceurl are unavilable (but there is at least one available endpoint), creating consumers and producers through PulsarClient will most likely fail. I think this is not as expected. If the code is robust enough, as long as there is ...
```
It's better to avoid references to other sources in the PIP motivation part. You might combine the "background knowledge" and "motivation" sections into a single section.

In the writing style it's better to avoid personal style, instead of "I think this is not expected", you could just make the claim which some rationale, for example "this is unexpected behavior for Pulsar users since when multiple endpoints are provided, it would be expected that when an endpoint is unavailable, an available endpoint would be used", etc.

lhotari commented on pip/pip-425.md:
```diff
@@ -0,0 +1,127 @@
+
+### Public API
+
+1. Add two optional configs `serviceUrlInitFaultQuarantineDuration` and `serviceUrlMaxFaultQuarantineDuration` in `ClientBuilder`
```
I don't think that `fault` is needed at all as part of the parameter name. `quarantine` already implies a fault.
`serviceUrlQuarantineInitDuration` and `serviceUrlQuarantineMaxDuration` would be easier to read and understand.

lhotari commented on pip/pip-425.md:
```diff
@@ -0,0 +1,127 @@
+    /**
+     * Configure the service URL init fault quarantine duration.
+     *
+     * <p>When the client is unable to connect to the service URL, it will wait for a certain amount of time before
+     * trying to recover the service URL endpoint. The init quarantine duration can be configured using this method.
```
this description isn't correct. There's not recovery happening. It's simply about quarantining the endpoint url for a specific duration that is determined in a certain emponential way. A successful usage of the endpoint url will reset the duration to the initial value and move it back to the available addresses pool.

lhotari reviewed (commented):
good progress! added some follow up comments

AuroraTwinkle commented on pip/pip-425.md:
```diff
@@ -0,0 +1,133 @@
+     * Configure the service URL init quarantine duration.
+     *
+     * <p>When the client is unable to connect to the a endpint from serviceUrl, that endpoint
+     *  will be quarantined for a specific duration that is determined in a certain
```
fixed

AuroraTwinkle commented:
> Hi @AuroraTwinkle
> 
> The proposal looks good to me. I’ve added a few comments to help clarify the problem and the proposed solution, making it easier to understand.

Ok, I will fix them later, Thanks!

AuroraTwinkle commented:
> Hi @AuroraTwinkle
> 
> The proposal looks good to me. I’ve added a few comments to help clarify the problem and the proposed solution, making it easier to understand.

@codelipenghui Very interesting and detailed suggestions, I have fixed them, thank you again!

315157973 commented on pip/pip-425.md:
````diff
@@ -0,0 +1,147 @@
+        // Subclass can override this method to implement host availability tracking
+    }
+```
+When resolve one address in serviceUrl, the `ConnectionPool` will invoke `markHostAvailability` to tell availability about that address, so the `ServiceNameResolver` can remove it from `availableAddressPool` if unavailable (or recover back to `availableAddressPool` if available). If all addressed in  `availableAddressPool` are unavailable, the resolver implementation could start getting random ad ...
````
It seems we need to add a circuit breaker mechanism. Can we simply address your problem by:
1. Changing from establishing connections by randomly selecting nodes to using a round-robin approach.
2. To prevent hotspots caused by multiple clients connecting to the same node , shuffling the list of IPs used for round-robin.

AuroraTwinkle replied:
OK, that was a typo, what I meant was to use a round-robin approach to select endpoint.

AuroraTwinkle replied:
https://github.com/apache/pulsar/blob/5841238d49695e29f557e22ce964d5b7470a354a/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarServiceNameResolver.java#L46-L58

@315157973 OKay, maybe I didn't explain it clearly enough.
This pip will not change the algorithm for selecting address, and will still maintain the original round-robin algorithm. What we change is the value of the list variable in the `resolveHost` function. When there are available addresses, a address are selected from the available address list. When there are no available addresses,  fall back to the original list (all address passed in by serviceUrl) for selection.

lhotari reviewed (approved):
LGTM, good work @AuroraTwinkle

liangyepianzhou commented:
@lhotari @codelipenghui @315157973 We need more votes to close this PIP. Could you please help vote when you have a moment?
https://lists.apache.org/thread/c2zvjwf7bqp8nc2rpzbxd4kdtztk23xp
